Improvements crazyflie

knmcguire opened the pull request:
**Description**
This is an improvement of the Bitcraze Crazyflie robot. The model has now a fasthelix, altitude control to keyboard controller. Also there is an python controller which features an autonomous mode, which is able to do wall following in a more complicated environment (the apartment) which is added as a second world.

**Related Issues**
This pull-request fixes issue #4711 partly (point 1-3 and 5) but the other issues are not mandatory and mostly improvements, so perhaps we can leave it at that and make the issue once we are ready to make a new PR

**Tasks**
These are the things done in this request
  - [x] Implemented fast helix
  - [x]  Altitude control to keyboard
  - [x] More complex world called crazyflie_apartment
  - [x]  New controller with autonomy

**Documentation**
No changes to documentation necessary.

**Screenshots**
This is a screenshot of the more complex behavior
![crazyflie_apartement_1](https://user-images.githubusercontent.com/11061260/234573294-54565646-e6e6-411c-bb90-012b73a70ac5.png)

knmcguire commented:
ahh a failed test. I have indeed reused the fast helix of the dji example and I would have gotten away with it too if it weren't for you meddling checks! :imp: (scoobydoo references)

I can make another one, no problem. I was just a bit lazy! :smile:

omichel commented:
You don't need to make another one, simply reusing the one of the dji is fine, but you should not copy the file.

omichel commented on projects/robots/bitcraze/crazyflie/protos/Crazyflie.proto:
```diff
@@ -30,153 +30,195 @@ PROTO Crazyflie [
+                baseColor 0.1 0.1 0.1
+                baseColorMap ImageTexture {
+                    url [
+                    "textures/fast_helix.png"
```
Assuming you re-used this one:
```suggestion
                    "webots://projects/robots/dji/mavic/protos/Mavic2Pro/textures/mavic_fast_helix.png"
```

knmcguire replied:
alright let's go for that then. I might make a mental note to fix this in the future as it would be nicer to have the meshes and textures separated from the proto without external dependencies, but it won't be urgent.

knmcguire commented:
> alright let's go for that then. I might make a mental note to fix this in the future as it would be nicer to have the meshes and textures separated from the proto without external dependencies, but it won't be urgent.

Or perhaps it can be a standard texture of some kind? Anyway, just a thought!

omichel commented:
> Or perhaps it can be a standard texture of some kind? Anyway, just a thought!

Yes, we could place a default/generic helix texture somewhere, like in `webots://projects/default/protos/textures/fast_helix.png` which could be used by both the Mavic and the Crazyflie.

omichel commented on projects/robots/bitcraze/crazyflie/protos/textures/fast_helix.png:
Obviously, this file should now be removed.

knmcguire commented:
I've also moved the fast helix proto to a more generic location now. Let's hope the CI now passes

knmcguire commented:
I'll fix the pep8 issues soon

knmcguire commented:
hmm... I think i need to format the c files with clang-9, however I only have clang-14 on my machine and clang-9 doesn't work on 22.04... so let me see what I can do

ygoumaz commented:
The sources tests use clang format 14 on all 3 platforms. See [tests_sources.yml](https://github.com/cyberbotics/webots/blob/develop/.github/workflows/tests_sources.yml) workflow file.

knmcguire commented:
hmmm oke! that is weird. I saw clang9 somewhere and it didn't accept my clang 14 format in the CI before. so that is why I assumed that it might have been clang9

wonder what went wrong...

knmcguire commented:
I revert it back to the original format in [39703a5](https://github.com/cyberbotics/webots/pull/6107/commits/39703a5c99601fdf783bb159c56e38665046653b), so I'll put it back to the commit before

ygoumaz commented:
It is not impossible that some configuration file of clang has an old name that was not updated. I added the `test_sources` label so that sources are tested on all platforms.

knmcguire commented:
finally, it passed! 

I do not recommend clang formatting by hand... that was a bit frustrating.

Wondering why the clang-format-14 on my ubuntu 2022.04 and in the vscode plugin was not able to make this pass... seems to be quite elementary, or perhaps there are some extra conditions that I wasn't aware sure. The clang-format I had was quite aggressive with shortening lines (as you can see [here](https://github.com/cyberbotics/webots/pull/6107/commits/46b4d15124a8a59af25e8401e868d4c7c7dd9d91#diff-f4e761a5f259046a147c0d04fae948c96471dadf22be1e52de98c4765b17b925)).

knmcguire commented:
Alright, I've tested out if the changes didn't affect the simulation and the demo. I've marked it ready for review

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie/crazyflie.c:
```diff
@@ -102,22 +104,25 @@ int main(int argc, char **argv) {
   printf(" The Crazyflie can be controlled from your keyboard!\n");
   printf(" All controllable movement is in body coordinates\n");
-  printf("- Use the up, back, right and left button to move in the horizontal plane\n");
+  printf("- Use the up, back, right and left button to move in the horizontal "
+         "plane\n");
```
This one should be fine for clang-format as we are extending the default line length from 80 to 128 characters.
```suggestion
  printf("- Use the up, back, right and left button to move in the horizontal plane\n");
```

knmcguire replied:
Done in [3d59860](https://github.com/cyberbotics/webots/pull/6107/commits/3d598602614f7f338678f46e922b14c7e43326e6)

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie/pid_controller.c:
```diff
@@ -41,6 +41,7 @@ float constrain(float value, const float minVal, const float maxVal) {
 
 double pastAltitudeError, pastPitchError, pastRollError, pastYawRateError;
 double pastVxError, pastVyError;
+double altIntergrator;
```
```suggestion
double altitudeIntegrator;
```

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie/pid_controller.c:
```diff
@@ -49,6 +50,7 @@ void init_pid_attitude_fixed_height_controller() {
   pastRollError = 0;
   pastVxError = 0;
   pastVyError = 0;
+  altIntergrator = 0;
```
```suggestion
  altitudeIntegrator = 0;
```

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie/pid_controller.c:
```diff
@@ -74,6 +76,10 @@ void pid_fixed_height_controller(actual_state_t actual_state, desired_state_t *d
   control_commands->altitude =
     gains_pid.kp_z * constrain(altitudeError, -1, 1) + gains_pid.kd_z * altitudeDerivativeError + gains_pid.ki_z;
+
+  altIntergrator += altitudeError * dt;
```
```suggestion
  altitudeIntegrator += altitudeError * dt;
```

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie/pid_controller.c:
```diff
@@ -74,6 +76,10 @@ void pid_fixed_height_controller(actual_state_t actual_state, desired_state_t *d
+
+  altIntergrator += altitudeError * dt;
+  control_commands->altitude = gains_pid.kp_z * constrain(altitudeError, -1, 1) + gains_pid.kd_z * altitudeDerivativeError +
+                               gains_pid.ki_z * altIntergrator + 48;
```
```suggestion
                               gains_pid.ki_z * altitudeIntegrator + 48;
```

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie_py_wallfollowing/crazyflie_py_wallfollowing.py:
```diff
@@ -0,0 +1,191 @@
+    print(" The Crazyflie can be controlled from your keyboard!\n")
+    print(" All controllable movement is in body coordinates\n")
+    print("- Use the up, back, right and left button to move in the horizontal plane\n")
+    print("- Use Q and E to rotate around yaw ")
```
```suggestion
    print("- Use Q and E to rotate around yaw\n")
```

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie_py_wallfollowing/crazyflie_py_wallfollowing.py:
```diff
@@ -0,0 +1,191 @@
+
+        height_desired += height_diff_desired * dt
+
+        cameraData = camera.getImage()
```
```suggestion
        camera_data = camera.getImage()
```

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie_py_wallfollowing/crazyflie_py_wallfollowing.py:
```diff
@@ -0,0 +1,191 @@
+    m4_motor.setVelocity(1)
+
+    # Initialize Sensors
+    imu = robot.getDevice("inertial unit")
```
It looks a bit inconsistent to have some sensor name with a space whereas all others use a underscore instead.

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie_py_wallfollowing/crazyflie_py_wallfollowing.py:
```diff
@@ -0,0 +1,191 @@
+    first_time = True
+
+    # Crazyflie velocity PID controller
+    PID_CF = pid_velocity_fixed_height_controller()
```
We can accept the PID abbreviation because it is a very common one. However, the CF abbreviation is not clear and should be either removed or expanded.
```suggestion
    PID_CF = pid_velocity_fixed_height_controller()
```

omichel reviewed (changes requested):
Thank you. It looks pretty good.
I suggest a few minor edits to make the code more readable and compliant with our standards.

omichel commented:
In general we try to avoid abbreviations as much as possible to make the code easier to maintain.

knmcguire commented:
@omichel I've addressed the changes, and I've also updated the wall following file as I was right in the middle of improving the [code quality of the wallfollowing of the 'real' crazyflie python library](https://github.com/bitcraze/crazyflie-lib-python/pull/397) itself. This also makes the state machine a lot easier the follow.

knmcguire commented:
ah wait I've missed a couple of comments in the code. Let me address those first

knmcguire commented:
alright, now all comments are addressed and resolved.

knmcguire commented:
So, the tests have passed, and I've just merged the latest. If that all goes well, then perhaps ready to merge?

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie/pid_controller.c:
```diff
@@ -41,6 +41,7 @@ float constrain(float value, const float minVal, const float maxVal) {
 
 double pastAltitudeError, pastPitchError, pastRollError, pastYawRateError;
 double pastVxError, pastVyError;
+double altitudeIntergrator;
```
There is still a typo here:
```suggestion
double altitudeIntegrator;
```

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie/pid_controller.c:
```diff
@@ -49,6 +50,7 @@ void init_pid_attitude_fixed_height_controller() {
   pastRollError = 0;
   pastVxError = 0;
   pastVyError = 0;
+  altitudeIntergrator = 0;
```
```suggestion
  altitudeIntegrator = 0;
```

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie/pid_controller.c:
```diff
@@ -74,6 +76,10 @@ void pid_fixed_height_controller(actual_state_t actual_state, desired_state_t *d
   control_commands->altitude =
     gains_pid.kp_z * constrain(altitudeError, -1, 1) + gains_pid.kd_z * altitudeDerivativeError + gains_pid.ki_z;
+
+  altitudeIntergrator += altitudeError * dt;
+  control_commands->altitude = gains_pid.kp_z * constrain(altitudeError, -1, 1) + gains_pid.kd_z * altitudeDerivativeError +
+                               gains_pid.ki_z * altitudeIntergrator + 48;
```
```suggestion
  altitudeIntegrator += altitudeError * dt;
  control_commands->altitude = gains_pid.kp_z * constrain(altitudeError, -1, 1) + gains_pid.kd_z * altitudeDerivativeError +
                               gains_pid.ki_z * altitudeIntegrator + 48;
```

omichel commented on projects/robots/bitcraze/crazyflie/controllers/crazyflie_py_wallfollowing/crazyflie_py_wallfollowing.py:
```diff
@@ -0,0 +1,206 @@
+        camera_data = camera.getImage()
+
+        # get range in meters
+        range_front_value = range_front.getValue()/1000
+        range_right_value = range_right.getValue()/1000
+        range_left_value = range_left.getValue()/1000
```
```suggestion
        range_front_value = range_front.getValue() / 1000
        range_right_value = range_right.getValue() / 1000
        range_left_value = range_left.getValue() / 1000
```

omichel reviewed (changes requested):
Thank you.,
We are almost there.

knmcguire commented:
thanks for fixing the typos! I'll merge it after the tests has passed
